include/directx/d3dx12.h

```diff
@@ -4037,6 +4038,277 @@ class CD3DX12_NODE_MASK_SUBOBJECT
     D3D12_NODE_MASK m_Desc;
 };
 
+//------------------------------------------------------------------------------------------------
+// Peter: Implementation for the new CheckFeatureSupport API
+
+#define FEATURE_SUPPORT_GET(RETTYPE,FEATURE,OPTION) \
+RETTYPE OPTION() const \
+{ \
+    return FEATURE.OPTION; \
+}
+
+#define FEATURE_SUPPORT_GET_NAME(RETTYPE,FEATURE,OPTION,NAME) \
+RETTYPE NAME() const \
+{\
+    return FEATURE.OPTION; \
+}
+
+#define FEATURE_SUPPORT_GET_NODE_INDEXED(RETTYPE,FEATURE,OPTION) \
+RETTYPE OPTION(UINT NodeIndex=0) const \
+{\
+    return FEATURE[NodeIndex].OPTION; \
+}
+
+#define INITIALIZE_MEMBER_DATA_CHECKED(FEATURE, MEMBER) \
+if (FAILED(result = m_pDevice->CheckFeatureSupport(FEATURE, &MEMBER, sizeof(MEMBER)))) { \
+    return result; \
+}
+
+class CD3DX12FeatureSupport
+{
+public:
+    CD3DX12FeatureSupport(ID3D12Device* pDevice)
+    : m_pDevice(pDevice)
+    {
+        m_hStatus = Init();
+    }
+
+    HRESULT Init()
+    {
+        // TODO: Think about hot to catch and report errors
+        HRESULT result;
+        
+        // Initialize static feature support data structures
+        INITIALIZE_MEMBER_DATA_CHECKED(D3D12_FEATURE_D3D12_OPTIONS, m_dOptions);
+        INITIALIZE_MEMBER_DATA_CHECKED(D3D12_FEATURE_GPU_VIRTUAL_ADDRESS_SUPPORT, m_dGPUVASupport);
+        m_dShaderModel.HighestShaderModel = D3D_HIGHEST_SHADER_MODEL; // TODO: What to do if highest model is inconsistent in the DLL and in the header?
+        INITIALIZE_MEMBER_DATA_CHECKED(D3D12_FEATURE_SHADER_MODEL, m_dShaderModel);
+        INITIALIZE_MEMBER_DATA_CHECKED(D3D12_FEATURE_D3D12_OPTIONS1, m_dOptions1);
+        m_dRootSignature.HighestVersion = D3D_HIGHEST_ROOT_SIGNATURE_VERSION;
+        INITIALIZE_MEMBER_DATA_CHECKED(D3D12_FEATURE_ROOT_SIGNATURE, m_dRootSignature);
+        INITIALIZE_MEMBER_DATA_CHECKED(D3D12_FEATURE_D3D12_OPTIONS2, m_dOptions2);
+        INITIALIZE_MEMBER_DATA_CHECKED(D3D12_FEATURE_SHADER_CACHE, m_dShaderCache);

```

Comment:
Instead of using an early-out error approach for some of these, e.g. the shader cache flags, we should more cleanly handle the case where they're not supported by the D3D runtime.

Reply:
I think CheckFeatureSupport won't return an error even if a certain feature exists but is not supported by the runtime. It will fail, however, if the runtime doesn't know about the feature at all, for example the runtime has no D3D12_FEATURE_SHADER_CACHE in its feature enum. I think this is the issue you're referring to? This is indeed one of the goals of this project (which I haven't achieved yet...). 

One question are the features ordered in incremental order, i.e., is it true that newer features will always have a higher feature enum assigned to them? And is it true that if a newer feature is supported, an older feature would always be supported? Here by supported I mean the D3D12 runtime can recognize the feature enum when checking it. This can impact how I'm going to resolve this issue.

Reply:
I modified the code such that unsupported feature structs will be initialized to default/invalid values instead of failing.